<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>2. QC of Target Data - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "2. QC of Target Data";
    var mkdocs_page_input_path = "target.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../base/">1. QC of Base Data</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">2. QC of Target Data</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#obtaining-the-target-data">Obtaining the target data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#qc-checklist-target-data">QC checklist: Target data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#sample-size"># Sample size</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#file-transfer"># File transfer</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#genome-build"># Genome build</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#standard-gwas-qc"># Standard GWAS QC</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#ambiguous-snps"># Ambiguous SNPs</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#mismatching-snps"># Mismatching SNPs</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#duplicate-snps"># Duplicate SNPs</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#sex-chromosomes"># Sex chromosomes</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#sample-overlap"># Sample overlap</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#relatedness"># Relatedness</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#generate-final-qced-target-data-file">Generate final QC'ed target data file</a>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../cal_prs/">3. Calculating and analysing PRS</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../prsice/">PRSice-2</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../ldpred/">LDpred-2</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../lassosum/">lassosum</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing PRS Results</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>2. QC of Target Data</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/target.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="qc-of-target-data">QC of Target Data<a class="headerlink" href="#qc-of-target-data" title="Permanent link">&para;</a></h1>
<h2 id="obtaining-the-target-data">Obtaining the target data<a class="headerlink" href="#obtaining-the-target-data" title="Permanent link">&para;</a></h2>
<p>Target data consist of individual-level genotype-phenotype data, usually generated within your lab/department/collaboration. For this tutorial, we have simulated some genotype-phenotype data using the 1000 Genomes Project European samples. 
You can download the data <a href="https://drive.google.com/file/d/1uhJR_3sn7RA8U5iYQbcmTp6vFdQiF4F2/view?usp=sharing">here</a></p>
<p>Unzip the data as follow:</p>
<pre class="highlight"><code class="language-bash">unzip EUR.zip</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You will need <code>plink</code> in this section, which can be download from <a href="https://www.cog-genomics.org/plink/1.9/">here</a>.</p>
<p>Install the program <code>plink</code> and include its location in your PATH directory, which allows us to use <code>plink</code> instead of <code>./plink</code> in the commands below. If PLINK is not in your PATH directory and is instead in your working directory, replace all instances of <code>plink</code> in the tutorial with <code>./plink</code>.</p>
</div>
<h2 id="qc-checklist-target-data">QC checklist: Target data<a class="headerlink" href="#qc-checklist-target-data" title="Permanent link">&para;</a></h2>
<p>Below are the QC steps that comprise the QC checklist for the target data.</p>
<h2 id="sample-size"># Sample size<a class="headerlink" href="#sample-size" title="Permanent link">&para;</a></h2>
<p>We recommend that users only perform PRS analyses on target data of at least 100 individuals. The sample size of our target data here is 503 individuals. </p>
<h2 id="file-transfer"># File transfer<a class="headerlink" href="#file-transfer" title="Permanent link">&para;</a></h2>
<p>Usually we do not need to download and transfer the target data file because it is typically generated locally. However, the file should contain an md5sum code in case we send the data file to collaborators who may want to confirm that the file has not changed during the transfer.</p>
<details class="note"><summary>What is the md5sum code for each of the target files?</summary><table>
<thead>
<tr>
<th align="center">File</th>
<th align="center">md5sum</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>EUR.bed</strong></td>
<td align="center">98bcef133f683b1272d3ea5f97742e0e</td>
</tr>
<tr>
<td align="center"><strong>EUR.bim</strong></td>
<td align="center">6b286002904880055a9c94e01522f059</td>
</tr>
<tr>
<td align="center"><strong>EUR.cov</strong></td>
<td align="center">85ed18288c708e095385418517e9c3bd</td>
</tr>
<tr>
<td align="center"><strong>EUR.fam</strong></td>
<td align="center">e7b856f0c7bcaffc8405926d08386e97</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">dd445ce969a81cded20da5c88b82d4df</td>
</tr>
</tbody>
</table>
</details>
<h2 id="genome-build"># Genome build<a class="headerlink" href="#genome-build" title="Permanent link">&para;</a></h2>
<p>As stated in the base data section, the genome build for our base and target data is the same, as it should be.</p>
<h2 id="standard-gwas-qc"># Standard GWAS QC<a class="headerlink" href="#standard-gwas-qc" title="Permanent link">&para;</a></h2>
<p>The target data must be quality controlled to at least the standards 
implemented in GWAS studies, e.g. removing SNPs with low genotyping rate, 
low minor allele frequency, out of Hardy-Weinberg Equilibrium, removing
individuals with low genotyping rate 
(see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/">Marees et al</a>).</p>
<p>The following <code>plink</code> command applies some of these QC metrics to the target data:</p>
<pre class="highlight"><code class="language-bash">plink \
    --bfile EUR \
    --maf 0.01 \
    --hwe 1e-6 \
    --geno 0.01 \
    --mind 0.01 \
    --write-snplist \
    --make-just-fam \
    --out EUR.QC</code></pre>
<p>Each of the parameters corresponds to the following</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">bfile</td>
<td align="center">EUR</td>
<td align="left">Informs <code>plink</code> that the input genotype files should have a prefix of <code>EUR</code></td>
</tr>
<tr>
<td align="center">maf</td>
<td align="center">0.01</td>
<td align="left">Removes all SNPs with minor allele frequency less than 0.05. Genotyping errors typically have a larger influence on SNPs with low MAF. Studies with large sample sizes could apply a lower MAF threshold</td>
</tr>
<tr>
<td align="center">hwe</td>
<td align="center">1e-6</td>
<td align="left">Removes SNPs with low P-value from the Hardy-Weinberg Equilibrium Fisher's exact or chi-squared test. SNPs with significant P-values from the HWE test are more likely affected by genotyping error or the effects of natural selection. Filtering should be performed on the control samples to avoid filtering SNPs that are causal (under selection in cases). When phenotype information is included, plink will automatically perform the filtering in the controls.</td>
</tr>
<tr>
<td align="center">geno</td>
<td align="center">0.01</td>
<td align="left">Excludes SNPs that are missing in a high fraction of subjects. A two-stage filtering process is usually performed (see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/">Marees et al</a>).</td>
</tr>
<tr>
<td align="center">mind</td>
<td align="center">0.01</td>
<td align="left">Excludes individuals who have a high rate of genotype missingness, since this may indicate problems in the DNA sample or processing. (see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/">Marees et al</a> for more details).</td>
</tr>
<tr>
<td align="center">make-just-fam</td>
<td align="center">-</td>
<td align="left">Informs <code>plink</code> to only generate the QC'ed sample name to avoid generating the .bed file.</td>
</tr>
<tr>
<td align="center">write-snplist</td>
<td align="center">-</td>
<td align="left">Informs <code>plink</code> to only generate the QC'ed SNP list to avoid generating the .bed file.</td>
</tr>
<tr>
<td align="center">out</td>
<td align="center">EUR.QC</td>
<td align="left">Informs <code>plink</code> that all output should have a prefix of <code>EUR.QC</code></td>
</tr>
</tbody>
</table>
<details class="note"><summary>How many SNPs and samples were filtered?</summary><ul>
<li><code>14</code> samples were removed due to a high rate of genotype missingness</li>
<li><code>5,353</code> SNP were removed due to missing genotype data</li>
<li><code>944</code> SNPs were removed due to being out of Hardy-Weinberg Equilibrium</li>
<li><code>5,061</code> SNPs were removed due to low minor allele frequency</li>
</ul>
</details>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Normally, we can generate a new genotype file using the new sample list.
However,  this will use up a lot of storage space. Using <code>plink</code>'s
<code>--extract</code>, <code>--exclude</code>, <code>--keep</code>, <code>--remove</code>, <code>--make-just-fam</code> and <code>--write-snplist</code> functions, we can work 
solely on the list of samples and SNPs without duplicating the 
genotype file, reducing the storage space usage.  </p>
</div>
<p>Very high or low heterozygosity rates in individuals could be due to DNA contamination or to high levels of inbreeding. Therefore, samples with extreme heterozygosity are typically removed prior to downstream analyses. </p>
<p>First, we perform prunning to remove highly correlated SNPs:</p>
<pre class="highlight"><code class="language-bash">plink \
    --bfile EUR \
    --keep EUR.QC.fam \
    --extract EUR.QC.snplist \
    --indep-pairwise 200 50 0.25 \
    --out EUR.QC</code></pre>
<p>Each of the parameters corresponds to the following</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">bfile</td>
<td align="center">EUR</td>
<td align="left">Informs <code>plink</code> that the input genotype files should have a prefix of <code>EUR</code></td>
</tr>
<tr>
<td align="center">keep</td>
<td align="center">EUR.QC.fam</td>
<td align="left">Informs <code>plink</code> that we only want to use samples in <code>EUR.QC.fam</code> in the analysis</td>
</tr>
<tr>
<td align="center">extract</td>
<td align="center">EUR.QC.snplist</td>
<td align="left">Informs <code>plink</code> that we only want to use SNPs in <code>EUR.QC.snplist</code> in the analysis</td>
</tr>
<tr>
<td align="center">indep-pairwise</td>
<td align="center">200 50 0.25</td>
<td align="left">Informs <code>plink</code> that we wish to perform prunning with a window size of 200 variants, sliding across the genome with step size of 50 variants at a time, and filter out any SNPs with LD <span class="arithmatex">\(r^2\)</span> higher than 0.25</td>
</tr>
<tr>
<td align="center">out</td>
<td align="center">EUR.QC</td>
<td align="left">Informs <code>plink</code> that all output should have a prefix of <code>EUR.QC</code></td>
</tr>
</tbody>
</table>
<p>This will generate two files 1) <strong>EUR.QC.prune.in</strong> and 2) <strong>EUR.QC.prune.out</strong>. All SNPs within <strong>EUR.QC.prune.in</strong> have a pairwise <span class="arithmatex">\(r^2 &lt; 0.25\)</span>. </p>
<p>Heterozygosity rates can then be computed using <code>plink</code>:</p>
<pre class="highlight"><code class="language-bash">plink \
    --bfile EUR \
    --extract EUR.QC.prune.in \
    --keep EUR.QC.fam \
    --het \
    --out EUR.QC</code></pre>
<p>This will generate the <strong>EUR.QC.het</strong> file, which contains F coefficient estimates for assessing heterozygosity.
We will remove individuals with F coefficients that are more than 3 standard deviation (SD) units from the mean, which can be performed using the following <code>R</code> command (assuming that you have R downloaded, then you can open an <code>R</code> session by typing <code>R</code> in your terminal):</p>
<div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Without library</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">dat &lt;- read.table("EUR.QC.het", header=T) # Read in the EUR.het file, specify it has header
m &lt;- mean(dat$F) # Calculate the mean  
s &lt;- sd(dat$F) # Calculate the SD
valid &lt;- subset(dat, F &lt;= m+3*s &amp; F &gt;= m-3*s) # Get any samples with F coefficient within 3 SD of the population mean
write.table(valid[,c(1,2)], "EUR.valid.sample", quote=F, row.names=F) # print FID and IID for valid samples
q() # exit R</code></pre>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">With data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(data.table)
# Read in file
dat &lt;- fread("EUR.QC.het")
# Get samples with F coefficient within 3 SD of the population mean
valid &lt;- dat[F&lt;=mean(F)+3*sd(F) &amp; F&gt;=mean(F)-3*sd(F)] 
# print FID and IID for valid samples
fwrite(valid[,c("FID","IID")], "EUR.valid.sample", sep="\t") 
q() # exit R</code></pre>
</div>
</div>
<details class="note"><summary>How many samples were excluded due to high heterozygosity rate?</summary><ul>
<li><code>2</code> samples were excluded</li>
</ul>
</details>
<h2 id="ambiguous-snps"># Ambiguous SNPs<a class="headerlink" href="#ambiguous-snps" title="Permanent link">&para;</a></h2>
<p>These were removed during the base data QC.</p>
<h2 id="mismatching-snps"># Mismatching SNPs<a class="headerlink" href="#mismatching-snps" title="Permanent link">&para;</a></h2>
<p>SNPs that have mismatching alleles reported in the base and target data may be resolvable by strand-flipping the alleles to their complementary alleles in e.g. the target data, such as for a SNP with A/C in the base data and G/T in the target. This can be achieved with the following steps:</p>
<p>1. Load the bim file, the summary statistic and the QC SNP list into R</p>
<div class="tabbed-set" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">Without data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Read in bim file
bim &lt;- read.table("EUR.bim")
colnames(bim) &lt;- c("CHR", "SNP", "CM", "BP", "B.A1", "B.A2")
# Read in QCed SNPs
qc &lt;- read.table("EUR.QC.snplist", header = F, stringsAsFactors = F)
# Read in the GWAS data
height &lt;-
    read.table(gzfile("Height.QC.gz"),
            header = T,
            stringsAsFactors = F, 
            sep="\t")
# Change all alleles to upper case for easy comparison
height$A1 &lt;- toupper(height$A1)
height$A2 &lt;- toupper(height$A2)
bim$B.A1 &lt;- toupper(bim$B.A1)
bim$B.A2 &lt;- toupper(bim$B.A2)</code></pre>
</div>
<input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><label for="__tabbed_2_2">With data.table and magrittr</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># magrittr allow us to do piping, which help to reduce the 
# amount of intermediate data types
library(data.table)
library(magrittr)
# Read in bim file 
bim &lt;- fread("EUR.bim") %&gt;%
    # Note: . represents the output from previous step
    # The syntax here means, setnames of the data read from
    # the bim file, and replace the original column names by 
    # the new names
    setnames(., colnames(.), c("CHR", "SNP", "CM", "BP", "B.A1", "B.A2")) %&gt;%
    # And immediately change the alleles to upper cases
    .[,c("B.A1","B.A2"):=list(toupper(B.A1), toupper(B.A2))]
# Read in summary statistic data (require data.table v1.12.0+)
height &lt;- fread("Height.QC.gz") %&gt;%
    # And immediately change the alleles to upper cases
    .[,c("A1","A2"):=list(toupper(A1), toupper(A2))]
# Read in QCed SNPs
qc &lt;- fread("EUR.QC.snplist", header=F)</code></pre>
</div>
</div>
<p>2. Identify SNPs that require strand flipping </p>
<div class="tabbed-set" data-tabs="3:2"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Without data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Merge summary statistic with target
info &lt;- merge(bim, height, by = c("SNP", "CHR", "BP"))
# Filter QCed SNPs
info &lt;- info[info$SNP %in% qc$V1,]
# Function for finding the complementary allele
complement &lt;- function(x) {
    switch (
        x,
        "A" = "T",
        "C" = "G",
        "T" = "A",
        "G" = "C",
        return(NA)
    )
}
# Get SNPs that have the same alleles across base and target
info.match &lt;- subset(info, A1 == B.A1 &amp; A2 == B.A2)
# Identify SNPs that are complementary between base and target
info$C.A1 &lt;- sapply(info$B.A1, complement)
info$C.A2 &lt;- sapply(info$B.A2, complement)
info.complement &lt;- subset(info, A1 == C.A1 &amp; A2 == C.A2)
# Update the complementary alleles in the bim file
# This allow us to match the allele in subsequent analysis
complement.snps &lt;- bim$SNP %in% info.complement$SNP
bim[complement.snps,]$B.A1 &lt;-
    sapply(bim[complement.snps,]$B.A1, complement)
bim[complement.snps,]$B.A2 &lt;-
    sapply(bim[complement.snps,]$B.A2, complement)</code></pre>
</div>
<input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><label for="__tabbed_3_2">With data.table and magrittr</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Merge summary statistic with target
info &lt;- merge(bim, height, by=c("SNP", "CHR", "BP")) %&gt;%
    # And filter out QCed SNPs
    .[SNP %in% qc[,V1]]

# Function for calculating the complementary allele
complement &lt;- function(x){
    switch (x,
        "A" = "T",
        "C" = "G",
        "T" = "A",
        "G" = "C",
        return(NA)
    )
} 
# Get SNPs that have the same alleles across base and target
info.match &lt;- info[A1 == B.A1 &amp; A2 == B.A2, SNP]
# Identify SNPs that are complementary between base and target
com.snps &lt;- info[sapply(B.A1, complement) == A1 &amp;
                    sapply(B.A2, complement) == A2, SNP]
# Now update the bim file
bim[SNP %in% com.snps, c("B.A1", "B.A2") :=
        list(sapply(B.A1, complement),
            sapply(B.A2, complement))]</code></pre>
</div>
</div>
<p>3. Identify SNPs that require recoding in the target (to ensure the coding allele in the target data is the effective allele in the base summary statistic)</p>
<div class="tabbed-set" data-tabs="4:2"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">Without data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># identify SNPs that need recoding
info.recode &lt;- subset(info, A1 == B.A2 &amp; A2 == B.A1)
# Update the recode SNPs
recode.snps &lt;- bim$SNP %in% info.recode$SNP
tmp &lt;- bim[recode.snps,]$B.A1
bim[recode.snps,]$B.A1 &lt;- bim[recode.snps,]$B.A2
bim[recode.snps,]$B.A2 &lt;- tmp

# identify SNPs that need recoding &amp; complement
info.crecode &lt;- subset(info, A1 == C.A2 &amp; A2 == C.A1)
# Update the recode + strand flip SNPs
com.snps &lt;- bim$SNP %in% info.crecode$SNP
tmp &lt;- bim[com.snps,]$B.A1
bim[com.snps,]$B.A1 &lt;- as.character(sapply(bim[com.snps,]$B.A2, complement))
bim[com.snps,]$B.A2 &lt;- as.character(sapply(tmp, complement))

# Output updated bim file
write.table(
    bim,
    "EUR.QC.adj.bim",
    quote = F,
    row.names = F,
    col.names = F,
    sep="\t"
)</code></pre>
</div>
<input id="__tabbed_4_2" name="__tabbed_4" type="radio" /><label for="__tabbed_4_2">With data.table  and magrittr</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># identify SNPs that need recoding
recode.snps &lt;- info[B.A1==A2 &amp; B.A2==A1, SNP]
# Update the bim file
bim[SNP %in% recode.snps, c("B.A1", "B.A2") :=
        list(B.A2, B.A1)]

# identify SNPs that need recoding &amp; complement
com.recode &lt;- info[sapply(B.A1, complement) == A2 &amp;
                    sapply(B.A2, complement) == A1, SNP]
# Now update the bim file
bim[SNP %in% com.recode, c("B.A1", "B.A2") :=
        list(sapply(B.A2, complement),
            sapply(B.A1, complement))]
# Write the updated bim file
fwrite(bim, "EUR.QC.adj.bim", col.names=F, sep="\t")</code></pre>
</div>
</div>
<p>4. Identify SNPs that have different allele in base and target (usually due to difference in genome build or Indel)</p>
<div class="tabbed-set" data-tabs="5:2"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">Without data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">mismatch &lt;-
    bim$SNP[!(bim$SNP %in% info.match$SNP |
                bim$SNP %in% info.complement$SNP | 
                bim$SNP %in% info.recode$SNP |
                bim$SNP %in% info.crecode$SNP)]
write.table(
    mismatch,
    "EUR.mismatch",
    quote = F,
    row.names = F,
    col.names = F
)
q() # exit R</code></pre>
</div>
<input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><label for="__tabbed_5_2">With data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">mismatch &lt;- bim[!(SNP %in% info.match |
                    SNP %in% com.snps |
                    SNP %in% recode.snps |
                    SNP %in% com.recode), SNP]
write.table(mismatch, "EUR.mismatch", quote=F, row.names=F, col.names=F)
q() # exit R</code></pre>
</div>
</div>
<p>5. Replace <strong>EUR.bim</strong> with <strong>EUR.QC.adj.bim</strong>:</p>
<p><pre class="highlight"><code class="language-bash"># Make a back up
mv EUR.bim EUR.bim.bk
ln -s EUR.QC.adj.bim EUR.bim</code></pre>
The above commands do the following:</p>
<ol>
<li>Rename <strong>EUR.bim</strong> to <strong>EUR.bim.bk</strong></li>
<li>Soft linking (<code>ln -s</code>) <strong>EUR.QC.adj.bim</strong> as <strong>EUR.bim</strong></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Most PRS software will perform strand-flipping automatically, thus this step is usually not required.</p>
</div>
<h2 id="duplicate-snps"># Duplicate SNPs<a class="headerlink" href="#duplicate-snps" title="Permanent link">&para;</a></h2>
<p>Make sure to remove any duplicate SNPs in your target data (these target data were simulated and so include no duplicated SNPs).</p>
<h2 id="sex-chromosomes"># Sex chromosomes<a class="headerlink" href="#sex-chromosomes" title="Permanent link">&para;</a></h2>
<p>Sometimes sample mislabelling can occur, which may lead to invalid results. One indication of a mislabelled sample is a difference between reported sex and that indicated by the sex chromosomes. While this may be due to a difference in sex and gender identity, it could also reflect mislabeling of samples or misreporting and, thus, individuals in which there is a mismatch between biological and reported sex are typically removed. A sex check can be performed in PLINK, in which individuals are called as females if their X chromosome homozygosity estimate (F statistic) is &lt; 0.2 and as males if the estimate is &gt; 0.8.</p>
<p>Before performing a sex check, pruning should be performed (see <a href="./#35-standard-gwas-qc">here</a>).
A sex check can then easily be conducted using <code>plink</code>
<pre class="highlight"><code class="language-bash">plink \
    --bfile EUR \
    --extract EUR.QC.prune.in \
    --keep EUR.valid.sample \
    --check-sex \
    --out EUR.QC</code></pre></p>
<p>This will generate a file called <strong>EUR.QC.sexcheck</strong> containing the F-statistics for each individual. Individuals are typically called as being biologically male if the F-statistic is &gt; 0.8 and biologically female if F &lt; 0.2.</p>
<div class="tabbed-set" data-tabs="6:2"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><label for="__tabbed_6_1">Without library</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Read in file
valid &lt;- read.table("EUR.valid.sample", header=T)
dat &lt;- read.table("EUR.QC.sexcheck", header=T)
valid &lt;- subset(dat, STATUS=="OK" &amp; FID %in% valid$FID)
write.table(valid[,c("FID", "IID")], "EUR.QC.valid", row.names=F, col.names=F, sep="\t", quote=F) 
q() # exit R</code></pre>
</div>
<input id="__tabbed_6_2" name="__tabbed_6" type="radio" /><label for="__tabbed_6_2">With data.table</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(data.table)
# Read in file
valid &lt;- fread("EUR.valid.sample")
dat &lt;- fread("EUR.QC.sexcheck")[FID%in%valid$FID]
fwrite(dat[STATUS=="OK",c("FID","IID")], "EUR.QC.valid", sep="\t") 
q() # exit R</code></pre>
</div>
</div>
<details class="note"><summary>How many samples were excluded due mismatched Sex information?</summary><ul>
<li><code>4</code> samples were excluded</li>
</ul>
</details>
<h2 id="sample-overlap"># Sample overlap<a class="headerlink" href="#sample-overlap" title="Permanent link">&para;</a></h2>
<p>Since the target data were simulated there are no overlapping samples between the base and target data here (see the relevant section of <a href="https://www.nature.com/articles/s41596-020-0353-1">the paper</a> for discussion of the importance of avoiding sample overlap). </p>
<h2 id="relatedness"># Relatedness<a class="headerlink" href="#relatedness" title="Permanent link">&para;</a></h2>
<p>Closely related individuals in the target data may lead to overfitted results, limiting the generalisability of the results. </p>
<p>Before calculating the relatedness, pruning should be performed (see <a href="./#35-standard-gwas-qc">here</a>).
Individuals that have a first or second degree relative in the sample (<span class="arithmatex">\(\hat{\pi} &gt; 0.125\)</span>) can be removed with the following command:</p>
<pre class="highlight"><code class="language-bash">plink \
    --bfile EUR \
    --extract EUR.QC.prune.in \
    --keep EUR.QC.valid \
    --rel-cutoff 0.125 \
    --out EUR.QC</code></pre>
<details class="note"><summary>How many related samples were excluded?</summary><ul>
<li><code>0</code> samples were excluded</li>
</ul>
</details>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A greedy algorithm is used to remove closely related individuals in a way that optimizes the size of the sample retained.                However, the algorithm is dependent on the random seed used, which can generate different results. Therefore, to reproduce
the same result, you will need to specify the same random seed. </p>
<p>PLINK's algorithm for removing related individuals does not account for the phenotype under study. 
To minimize the removal of cases of a disease, the following algorithm can be used instead: 
<a href="https://gitlab.com/choishingwan/GreedyRelated">GreedyRelated</a>.</p>
</div>
<h2 id="generate-final-qced-target-data-file">Generate final QC'ed target data file<a class="headerlink" href="#generate-final-qced-target-data-file" title="Permanent link">&para;</a></h2>
<p>After performing the full analysis, you can generate a QC'ed data set with the following command:
<pre class="highlight"><code class="language-bash">plink \
    --bfile EUR \
    --make-bed \
    --keep EUR.QC.rel.id \
    --out EUR.QC \
    --extract EUR.QC.snplist \
    --exclude EUR.mismatch</code></pre></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../cal_prs/" class="btn btn-neutral float-right" title="3. Calculating and analysing PRS">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../base/" class="btn btn-neutral" title="1. QC of Base Data"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../base/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../cal_prs/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
