<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>2. Target Genotype QC - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "2. Target Genotype QC";
    var mkdocs_page_input_path = "target.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../base/">1. QC of Summary Statistics</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">2. Target Genotype QC</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#obtaining-the-target-data">Obtaining the target data</a></li>
    

    <li class="toctree-l2"><a href="#qc-checklist-target-data">QC checklist: Target data</a></li>
    

    <li class="toctree-l2"><a href="#35-sample-size"># Sample size</a></li>
    

    <li class="toctree-l2"><a href="#35-file-transfer"># File transfer</a></li>
    

    <li class="toctree-l2"><a href="#35-genome-build"># Genome build</a></li>
    

    <li class="toctree-l2"><a href="#35-standard-gwas-qc"># Standard GWAS QC</a></li>
    

    <li class="toctree-l2"><a href="#35-ambiguous-snps"># Ambiguous SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-sex-chromosomes"># Sex chromosomes</a></li>
    

    <li class="toctree-l2"><a href="#35-mismatching-genotypes"># Mismatching genotypes</a></li>
    

    <li class="toctree-l2"><a href="#35-duplicate-snps"># Duplicate SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-sample-overlap"># Sample overlap</a></li>
    

    <li class="toctree-l2"><a href="#35-relatedness"># Relatedness</a></li>
    

    <li class="toctree-l2"><a href="#generate-final-qced-target-data-file">Generate final QC'ed target data file</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Running PRS</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../plink/">PLINK</a>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class="">
                    
    <a class="" href="../ldpred/">LDpred</a>
                </li>
                <li class="">
                    
    <a class="" href="../lassosum/">lassosum</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../plink_visual/">4. Visualizing PRS</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>2. Target Genotype QC</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/target.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="obtaining-the-target-data">Obtaining the target data<a class="headerlink" href="#obtaining-the-target-data" title="Permanent link">&para;</a></h1>
<p>Target data consist of individual-level genotype-phenotype data, usually generated within your lab/department/collaboration. For this tutorial, we have simulated some genotype-phenotype data using the 1000 Genomes Project European samples. 
You can download the data <a href="https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/EUR.zip">here</a> or you can download the data using the following command:</p>
<div class="codehilite"><pre><span></span>curl https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/EUR.zip -L -O
</pre></div>

<p>Unzip the data as follow:</p>
<div class="codehilite"><pre><span></span>unzip EUR.zip
</pre></div>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Install the program PLINK and include its location in your PATH directory, which allows us to use <code>plink</code> instead of <code>./plink</code> in the commands below. If PLINK is not in your PATH directory and is instead in your working directory, then replace all instances of <code>plink</code> in the tutorial with <code>./plink</code>.</p>
</div>
<h1 id="qc-checklist-target-data">QC checklist: Target data<a class="headerlink" href="#qc-checklist-target-data" title="Permanent link">&para;</a></h1>
<p>Below are the QC steps that comprise the QC checklist for the target data.</p>
<h1 id="35-sample-size"># Sample size<a class="headerlink" href="#35-sample-size" title="Permanent link">&para;</a></h1>
<p>We recommend that users only perform PRS analyses on target data of at least 100 individuals. The sample size of our target data here is 503 individuals. </p>
<h1 id="35-file-transfer"># File transfer<a class="headerlink" href="#35-file-transfer" title="Permanent link">&para;</a></h1>
<p>Usually we do not need to download and transfer the target data file because it is typically generated locally. However, the file should contain an md5sum code in case we send the data file to collaborators who may want to confirm that the file has not changed during the transfer.</p>
<details class="note"><summary>What is the md5sum code for each of the target files?</summary><table>
<thead>
<tr>
<th align="center">File</th>
<th align="center">md5sum</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>EUR.bed</strong></td>
<td align="center">940f5a760b41270662eba6264b262a2d</td>
</tr>
<tr>
<td align="center"><strong>EUR.bim</strong></td>
<td align="center">a528020cc2448aa04a7499f13bf9f16a</td>
</tr>
<tr>
<td align="center"><strong>EUR.covariate</strong></td>
<td align="center">afff13f8f9e15815f2237a62b8bec00b</td>
</tr>
<tr>
<td align="center"><strong>EUR.fam</strong></td>
<td align="center">17e8184fb03c690db6980bb7499d4982</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">052beb4cae32ac7673f1d6b9e854c85b</td>
</tr>
</tbody>
</table>
</details>
<h1 id="35-genome-build"># Genome build<a class="headerlink" href="#35-genome-build" title="Permanent link">&para;</a></h1>
<p>As stated in the base data section, the genome build for our base and target data is the same, as it should be.</p>
<h1 id="35-standard-gwas-qc"># Standard GWAS QC<a class="headerlink" href="#35-standard-gwas-qc" title="Permanent link">&para;</a></h1>
<p>The target data must be quality controlled to at least the standards 
implemented in GWAS studies, e.g. removing SNPs with low genotyping rate, 
low minor allele frequency, out of Hardy-Weinberg Equilibrium, removing
individuals with low genotyping rate 
(see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/">Marees et al</a>).</p>
<p>The following <code>plink</code> command applies some of these QC metrics to the target data:</p>
<p><div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR <span class="se">\</span>
    --maf <span class="m">0</span>.05 <span class="se">\</span>
    --hwe 1e-6 <span class="se">\</span>
    --geno <span class="m">0</span>.01 <span class="se">\</span>
    --mind <span class="m">0</span>.01 <span class="se">\</span>
    --write-snplist <span class="se">\</span>
    --make-just-fam <span class="se">\</span>
    --out EUR.QC
</pre></div>
Each of the parameters corresponds to the following</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">bfile</td>
<td align="center">EUR</td>
<td align="left">Informs <code>plink</code> that the input genotype files should have a prefix of <code>EUR</code></td>
</tr>
<tr>
<td align="center">maf</td>
<td align="center">0.05</td>
<td align="left">Removes all SNPs with minor allele frequency less than 0.05. Genotyping errors typically have a larger influence on SNPs with low MAF. Studies with large sample sizes could apply a lower MAF threshold</td>
</tr>
<tr>
<td align="center">hwe</td>
<td align="center">1e-6</td>
<td align="left">Removes SNPs with low P-value from the Hardy-Weinberg Equilibrium Fisher's exact or chi-squared test. SNPs with significant P-values from the HWE test are more likely affected by genotyping error or the effects of natural selection. Filtering should be performed on the control samples to avoid filtering SNPs that are causal (under selection in cases)</td>
</tr>
<tr>
<td align="center">geno</td>
<td align="center">0.01</td>
<td align="left">Excludes SNPs that are missing in a high fraction of subjects. A two-stage filtering process is usually performed (see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/">Marees et al</a>).</td>
</tr>
<tr>
<td align="center">mind</td>
<td align="center">0.01</td>
<td align="left">Excludes individuals who have a high rate of genotype missingness, since this may indicate problems in the DNA sample or processing. (see <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/">Marees et al</a> for more details).</td>
</tr>
<tr>
<td align="center">make-just-fam</td>
<td align="center">-</td>
<td align="left">Informs <code>plink</code> to only generate the QC'ed sample name to avoid generating the .bed file.</td>
</tr>
<tr>
<td align="center">write-snplist</td>
<td align="center">-</td>
<td align="left">Informs <code>plink</code> to only generate the QC'ed SNP list to avoid generating the .bed file.</td>
</tr>
<tr>
<td align="center">out</td>
<td align="center">EUR</td>
<td align="left">Informs <code>plink</code> that all output should have a prefix of <code>EUR</code></td>
</tr>
</tbody>
</table>
<details class="note"><summary>How many SNPs and samples were filtered?</summary><ul>
<li><code>5</code> samples were removed due to a high rate of genotype missingness</li>
<li><code>1</code> SNP were removed due to missing genotype data</li>
<li><code>872</code> SNPs were removed due to being out of Hardy-Weinberg Equilibrium</li>
<li><code>242,459</code> SNPs were removed due to low minor allele frequency</li>
</ul>
</details>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Normally, we can generate a new genotype file using the new sample list.
However,  this will use up a lot of storage space. Using <code>plink</code>'s
<code>--extract</code>, <code>--exclude</code>, <code>--keep</code>, <code>--remove</code>, <code>--make-just-fam</code> and <code>--write-snplist</code> functions, we can work 
solely on the list of samples and SNPs without duplicating the 
genotype file, reducing the storage space usage.  </p>
</div>
<p>Very high or low heterozygosity rates in individuals could be due to DNA contamination or to high levels of inbreeding. Therefore, samples with extreme heterozygosity are typically removed prior to downstream analyses. </p>
<p>First, we perform prunning to remove highly correlated SNPs:
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR <span class="se">\</span>
    --keep EUR.QC.fam <span class="se">\</span>
    --extract EUR.QC.snplist <span class="se">\</span>
    --indep-pairwise <span class="m">200</span> <span class="m">50</span> <span class="m">0</span>.25 <span class="se">\</span>
    --out EUR.QC
</pre></div></p>
<p>This will generate two files 1) <strong>EUR.QC.prune.in</strong> and 2) <strong>EUR.QC.prune.out</strong>.All SNPs within <strong>EUR.QC.prune.in</strong> have a pairwise <span class="arithmatex">\(r^2 &lt; 0.25\)</span>. </p>
<p>Heterozygosity rates can then be computed using <code>plink</code>:
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR <span class="se">\</span>
    --extract EUR.QC.prune.in <span class="se">\</span>
    --keep EUR.QC.fam <span class="se">\</span>
    --het <span class="se">\</span>
    --out EUR.QC
</pre></div></p>
<p>This will generate the <strong>EUR.QC.het</strong> file, which contains F coefficient estimates for assessing heterozygosity.
We will remove individuals with F coefficients that are more than 3 standard deviation (SD) units from the mean, which can be performed using the following <code>R</code> command (assuming that you have R downloaded, then you can open an <code>R</code> session by typing <code>R</code> in your terminal):</p>
<div class="superfences-tabs">
<input name="__tabs_1" type="radio" id="__tab_1_0" checked="checked" />
<label for="__tab_1_0">Without library</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>dat <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.QC.het&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span> <span class="c1"># Read in the EUR.het file, specify it has header</span>
m <span class="o">&lt;-</span> <span class="kp">mean</span><span class="p">(</span>dat<span class="o">$</span><span class="bp">F</span><span class="p">)</span> <span class="c1"># Calculate the mean  </span>
s <span class="o">&lt;-</span> sd<span class="p">(</span>dat<span class="o">$</span><span class="bp">F</span><span class="p">)</span> <span class="c1"># Calculate the SD</span>
valid <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>dat<span class="p">,</span> <span class="bp">F</span> <span class="o">&lt;=</span> m<span class="m">+3</span><span class="o">*</span>s <span class="o">&amp;</span> <span class="bp">F</span> <span class="o">&gt;=</span> m<span class="m">-3</span><span class="o">*</span>s<span class="p">)</span> <span class="c1"># Get any samples with F coefficient within 3 SD of the population mean</span>
write.table<span class="p">(</span>valid<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">)],</span> <span class="s">&quot;EUR.valid.sample&quot;</span><span class="p">,</span> quote<span class="o">=</span><span class="bp">F</span><span class="p">,</span> row.names<span class="o">=</span><span class="bp">F</span><span class="p">)</span> <span class="c1"># print FID and IID for valid samples</span>
</pre></div></div>
<input name="__tabs_1" type="radio" id="__tab_1_1" />
<label for="__tab_1_1">With data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
<span class="c1"># Read in file</span>
dat <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.QC.het&quot;</span><span class="p">)</span>
<span class="c1"># Get samples with F coefficient within 3 SD of the population mean</span>
valid <span class="o">&lt;-</span> dat<span class="p">[</span><span class="bp">F</span><span class="o">&lt;=</span><span class="kp">mean</span><span class="p">(</span><span class="bp">F</span><span class="p">)</span><span class="m">+3</span><span class="o">*</span>sd<span class="p">(</span><span class="bp">F</span><span class="p">)</span> <span class="o">&amp;</span> <span class="bp">F</span><span class="o">&gt;=</span><span class="kp">mean</span><span class="p">(</span><span class="bp">F</span><span class="p">)</span><span class="m">-3</span><span class="o">*</span>sd<span class="p">(</span><span class="bp">F</span><span class="p">)]</span> 
<span class="c1"># print FID and IID for valid samples</span>
fwrite<span class="p">(</span>valid<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)],</span> <span class="s">&quot;EUR.valid.sample&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span> 
</pre></div></div>
</div>
<details class="note"><summary>How many samples were excluded due to high heterozygosity rate?</summary><ul>
<li><code>7</code> samples were excluded</li>
</ul>
</details>
<h1 id="35-ambiguous-snps"># Ambiguous SNPs<a class="headerlink" href="#35-ambiguous-snps" title="Permanent link">&para;</a></h1>
<p>There were removed during the base data QC</p>
<h1 id="35-sex-chromosomes"># Sex chromosomes<a class="headerlink" href="#35-sex-chromosomes" title="Permanent link">&para;</a></h1>
<p>Sometimes sample mislabelling can occur, which may lead to invalid results. 
A good indication of a mislabelled sample is a mismatch between biological sex and reported sex. 
If the biological sex does not match up with the reported sex, then the sample may have been mislabelled.</p>
<p>Before performing a sex check, pruning should be performed (see <a href="./#35-standard-gwas-qc">here</a>).
A sex check can then easily be conducted using <code>plink</code>
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR <span class="se">\</span>
    --extract EUR.QC.prune.in <span class="se">\</span>
    --keep EUR.valid.sample <span class="se">\</span>
    --check-sex <span class="se">\</span>
    --out EUR.QC
</pre></div></p>
<p>This will generate a file called <strong>EUR.QC.sexcheck</strong> containing the F-statistics for each individual. Individuals are typically called as being biologically male if the F-statistic is &gt; 0.8 and biologically female if F &lt; 0.2.</p>
<div class="superfences-tabs">
<input name="__tabs_2" type="radio" id="__tab_2_0" checked="checked" />
<label for="__tab_2_0">Without library</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="c1"># Read in file</span>
valid <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.valid.sample&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
dat <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.QC.sexcheck&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
valid <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>dat<span class="p">,</span> STATUS<span class="o">==</span><span class="s">&quot;OK&quot;</span> <span class="o">&amp;</span> FID <span class="o">%in%</span> valid<span class="o">$</span>FID<span class="p">)</span>
write.table<span class="p">(</span>valid<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">)],</span> <span class="s">&quot;EUR.QC.valid&quot;</span><span class="p">,</span> row.names<span class="o">=</span><span class="bp">F</span><span class="p">,</span> col.names<span class="o">=</span><span class="bp">F</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> quote<span class="o">=</span><span class="bp">F</span><span class="p">)</span> 
</pre></div></div>
<input name="__tabs_2" type="radio" id="__tab_2_1" />
<label for="__tab_2_1">With data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
<span class="c1"># Read in file</span>
valid <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.valid.sample&quot;</span><span class="p">)</span>
dat <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.QC.sexcheck&quot;</span><span class="p">)[</span>FID<span class="o">%in%</span>valid<span class="o">$</span>FID<span class="p">]</span>
fwrite<span class="p">(</span>dat<span class="p">[</span>STATUS<span class="o">==</span><span class="s">&quot;OK&quot;</span><span class="p">,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)],</span> <span class="s">&quot;EUR.QC.valid&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span> 
</pre></div></div>
</div>
<details class="note"><summary>How many samples were excluded due mismatched Sex information?</summary><ul>
<li><code>2</code> samples were excluded</li>
</ul>
</details>
<h1 id="35-mismatching-genotypes"># Mismatching genotypes<a class="headerlink" href="#35-mismatching-genotypes" title="Permanent link">&para;</a></h1>
<p>In addition, when there are non-ambiguous mismatches in allele 
coding between the data sets, such as A/C in the base
and G/T in the target data, then this can be resolved by 
‘flipping’ the alleles in the target data to their complementary alleles. 
This can be achieved with the following steps: </p>
<p>1. Get the correct A1 alleles for the bim file</p>
<div class="superfences-tabs">
<input name="__tabs_3" type="radio" id="__tab_3_0" checked="checked" />
<label for="__tab_3_0">Without data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>bim <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.bim&quot;</span><span class="p">)</span>
<span class="kp">colnames</span><span class="p">(</span>bim<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;CHR&quot;</span><span class="p">,</span> <span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;CM&quot;</span><span class="p">,</span> <span class="s">&quot;BP&quot;</span><span class="p">,</span> <span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span>
qc <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.QC.snplist&quot;</span><span class="p">,</span> header <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> stringsAsFactors <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
height <span class="o">&lt;-</span>
    read.table<span class="p">(</span><span class="kp">gzfile</span><span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">),</span>
               header <span class="o">=</span> <span class="bp">T</span><span class="p">,</span>
               stringsAsFactors <span class="o">=</span> <span class="bp">F</span><span class="p">,</span> 
               sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="c1"># Change all alleles to upper case for easy comparison</span>
height<span class="o">$</span>A1 <span class="o">&lt;-</span> <span class="kp">toupper</span><span class="p">(</span>height<span class="o">$</span>A1<span class="p">)</span>
height<span class="o">$</span>A2 <span class="o">&lt;-</span> <span class="kp">toupper</span><span class="p">(</span>height<span class="o">$</span>A2<span class="p">)</span>
bim<span class="o">$</span>B.A1 <span class="o">&lt;-</span> <span class="kp">toupper</span><span class="p">(</span>bim<span class="o">$</span>B.A1<span class="p">)</span>
bim<span class="o">$</span>B.A2 <span class="o">&lt;-</span> <span class="kp">toupper</span><span class="p">(</span>bim<span class="o">$</span>B.A2<span class="p">)</span>
info <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>bim<span class="p">,</span> height<span class="p">,</span> by <span class="o">=</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;CHR&quot;</span><span class="p">,</span> <span class="s">&quot;BP&quot;</span><span class="p">))</span>
info <span class="o">&lt;-</span> info<span class="p">[</span>info<span class="o">$</span>SNP <span class="o">%in%</span> qc<span class="o">$</span>V1<span class="p">,]</span>
<span class="c1"># Function for finding the complementary allele</span>
complement <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">)</span> <span class="p">{</span>
    <span class="kr">switch</span> <span class="p">(</span>
        x<span class="p">,</span>
        <span class="s">&quot;A&quot;</span> <span class="o">=</span> <span class="s">&quot;T&quot;</span><span class="p">,</span>
        <span class="s">&quot;C&quot;</span> <span class="o">=</span> <span class="s">&quot;G&quot;</span><span class="p">,</span>
        <span class="s">&quot;T&quot;</span> <span class="o">=</span> <span class="s">&quot;A&quot;</span><span class="p">,</span>
        <span class="s">&quot;G&quot;</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">,</span>
        <span class="kr">return</span><span class="p">(</span><span class="kc">NA</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">}</span>
<span class="c1"># Get SNPs that have the same alleles across base and target</span>
info.match <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>info<span class="p">,</span> A1 <span class="o">==</span> B.A1 <span class="o">&amp;</span> A2 <span class="o">==</span> B.A2<span class="p">)</span>
<span class="c1"># Identify SNPs that are complementary between base and target</span>
info<span class="o">$</span>C.A1 <span class="o">&lt;-</span> <span class="kp">sapply</span><span class="p">(</span>info<span class="o">$</span>B.A1<span class="p">,</span> complement<span class="p">)</span>
info<span class="o">$</span>C.A2 <span class="o">&lt;-</span> <span class="kp">sapply</span><span class="p">(</span>info<span class="o">$</span>B.A2<span class="p">,</span> complement<span class="p">)</span>
info.complement <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>info<span class="p">,</span> A1 <span class="o">==</span> C.A1 <span class="o">&amp;</span> A2 <span class="o">==</span> C.A2<span class="p">)</span>
<span class="c1"># Update these allele coding in the bim file</span>
bim<span class="p">[</span>bim<span class="o">$</span>SNP <span class="o">%in%</span> info.complement<span class="o">$</span>SNP<span class="p">,]</span><span class="o">$</span>B.A1 <span class="o">&lt;-</span>
    <span class="kp">sapply</span><span class="p">(</span>bim<span class="p">[</span>bim<span class="o">$</span>SNP <span class="o">%in%</span> info.complement<span class="o">$</span>SNP<span class="p">,]</span><span class="o">$</span>B.A1<span class="p">,</span> complement<span class="p">)</span>
bim<span class="p">[</span>bim<span class="o">$</span>SNP <span class="o">%in%</span> info.complement<span class="o">$</span>SNP<span class="p">,]</span><span class="o">$</span>B.A2 <span class="o">&lt;-</span>
    <span class="kp">sapply</span><span class="p">(</span>bim<span class="p">[</span>bim<span class="o">$</span>SNP <span class="o">%in%</span> info.complement<span class="o">$</span>SNP<span class="p">,]</span><span class="o">$</span>B.A2<span class="p">,</span> complement<span class="p">)</span>
<span class="c1"># identify SNPs that need flipping</span>
info.flip <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>info<span class="p">,</span> A1 <span class="o">==</span> B.A2 <span class="o">&amp;</span> A2 <span class="o">==</span> B.A1<span class="p">)</span>
<span class="c1"># identify SNPs that need flipping &amp; complement</span>
info.cflip <span class="o">&lt;-</span> <span class="kp">subset</span><span class="p">(</span>info<span class="p">,</span> A1 <span class="o">==</span> C.A2 <span class="o">&amp;</span> A2 <span class="o">==</span> C.A1<span class="p">)</span>
<span class="c1"># Update these allele coding in the bim file</span>
<span class="c1"># Note that we don&#39;t flip the SNP here. We will let PLINK do that. </span>
com.snps <span class="o">&lt;-</span> bim<span class="o">$</span>SNP <span class="o">%in%</span> info.cflip<span class="o">$</span>SNP
bim<span class="p">[</span>com.snps<span class="p">,]</span><span class="o">$</span>B.A1 <span class="o">&lt;-</span> <span class="kp">sapply</span><span class="p">(</span>bim<span class="p">[</span>com.snps<span class="p">,]</span><span class="o">$</span>B.A1<span class="p">,</span> complement<span class="p">)</span>
bim<span class="p">[</span>com.snps<span class="p">,]</span><span class="o">$</span>B.A2 <span class="o">&lt;-</span> <span class="kp">sapply</span><span class="p">(</span>bim<span class="p">[</span>com.snps<span class="p">,]</span><span class="o">$</span>B.A2<span class="p">,</span> complement<span class="p">)</span>
<span class="c1"># Get list of SNPs that need to change the A1 encoding</span>
flip <span class="o">&lt;-</span> <span class="kp">rbind</span><span class="p">(</span>info.flip<span class="p">,</span> info.cflip<span class="p">)</span>
flip.snp <span class="o">&lt;-</span> <span class="kt">data.frame</span><span class="p">(</span>SNP <span class="o">=</span> flip<span class="o">$</span>SNP<span class="p">,</span> A1 <span class="o">=</span> flip<span class="o">$</span>A1<span class="p">)</span>
write.table<span class="p">(</span>flip.snp<span class="p">,</span>
            <span class="s">&quot;EUR.update.a1&quot;</span><span class="p">,</span>
            quote <span class="o">=</span> <span class="bp">F</span><span class="p">,</span>
            row.names <span class="o">=</span> <span class="bp">F</span><span class="p">)</span>
write.table<span class="p">(</span>
    bim<span class="p">,</span>
    <span class="s">&quot;EUR.QC.adj.bim&quot;</span><span class="p">,</span>
    quote <span class="o">=</span> <span class="bp">F</span><span class="p">,</span>
    row.names <span class="o">=</span> <span class="bp">F</span><span class="p">,</span>
    col.names <span class="o">=</span> <span class="bp">F</span>
<span class="p">)</span>
<span class="c1"># And we want to remove any SNPs that do not match with the base data</span>
mismatch <span class="o">&lt;-</span>
    bim<span class="o">$</span>SNP<span class="p">[</span><span class="o">!</span><span class="p">(</span>bim<span class="o">$</span>SNP <span class="o">%in%</span> info.match<span class="o">$</span>SNP <span class="o">|</span>
                  bim<span class="o">$</span>SNP <span class="o">%in%</span> info.complement<span class="o">$</span>SNP <span class="o">|</span> 
                  bim<span class="o">$</span>SNP <span class="o">%in%</span> flip<span class="o">$</span>SNP<span class="p">)]</span>
write.table<span class="p">(</span>
    mismatch<span class="p">,</span>
    <span class="s">&quot;EUR.mismatch&quot;</span><span class="p">,</span>
    quote <span class="o">=</span> <span class="bp">F</span><span class="p">,</span>
    row.names <span class="o">=</span> <span class="bp">F</span><span class="p">,</span>
    col.names <span class="o">=</span> <span class="bp">F</span>
<span class="p">)</span>
</pre></div></div>
<input name="__tabs_3" type="radio" id="__tab_3_1" />
<label for="__tab_3_1">With data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
bim <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.bim&quot;</span><span class="p">)</span>
bim.col <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;CHR&quot;</span><span class="p">,</span> <span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;CM&quot;</span><span class="p">,</span> <span class="s">&quot;BP&quot;</span><span class="p">,</span> <span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span>
setnames<span class="p">(</span>bim<span class="p">,</span> <span class="kp">colnames</span><span class="p">(</span>bim<span class="p">),</span> bim.col<span class="p">)</span>
qc <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;EUR.QC.snplist&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
height <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">)</span>
<span class="c1"># Change all alleles to upper case for easy comparison</span>
height<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;A1&quot;</span><span class="p">,</span><span class="s">&quot;A2&quot;</span><span class="p">)</span><span class="o">:=</span><span class="kt">list</span><span class="p">(</span><span class="kp">toupper</span><span class="p">(</span>A1<span class="p">),</span> <span class="kp">toupper</span><span class="p">(</span>A2<span class="p">))]</span>
bim<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span><span class="s">&quot;B.A2&quot;</span><span class="p">)</span><span class="o">:=</span><span class="kt">list</span><span class="p">(</span><span class="kp">toupper</span><span class="p">(</span>B.A1<span class="p">),</span> <span class="kp">toupper</span><span class="p">(</span>B.A2<span class="p">))]</span>
info <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>bim<span class="p">,</span> height<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;CHR&quot;</span><span class="p">,</span> <span class="s">&quot;BP&quot;</span><span class="p">))</span>
info <span class="o">&lt;-</span> info<span class="p">[</span>SNP <span class="o">%in%</span> qc<span class="o">$</span>V1<span class="p">]</span>
<span class="c1"># Function for calculating the complementary allele</span>
complement <span class="o">&lt;-</span> <span class="kr">function</span><span class="p">(</span>x<span class="p">){</span>
    <span class="kr">switch</span> <span class="p">(</span>x<span class="p">,</span>
        <span class="s">&quot;A&quot;</span> <span class="o">=</span> <span class="s">&quot;T&quot;</span><span class="p">,</span>
        <span class="s">&quot;C&quot;</span> <span class="o">=</span> <span class="s">&quot;G&quot;</span><span class="p">,</span>
        <span class="s">&quot;T&quot;</span> <span class="o">=</span> <span class="s">&quot;A&quot;</span><span class="p">,</span>
        <span class="s">&quot;G&quot;</span> <span class="o">=</span> <span class="s">&quot;C&quot;</span><span class="p">,</span>
        <span class="kr">return</span><span class="p">(</span><span class="kc">NA</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">}</span>
<span class="c1"># Identify SNPs that are complementary between base and target</span>
com.snps <span class="o">&lt;-</span> info<span class="p">[</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A1 <span class="o">&amp;</span>
                     <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A2<span class="p">,</span> SNP<span class="p">]</span>
<span class="c1"># Now update the bim file</span>
bim<span class="p">[</span>SNP <span class="o">%in%</span> com.snps<span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span> <span class="o">:=</span>
        <span class="kt">list</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">),</span>
             <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))]</span>
<span class="c1"># And update the info structure</span>
info<span class="p">[</span>SNP <span class="o">%in%</span> com.snps<span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span> <span class="o">:=</span>
        <span class="kt">list</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">),</span>
             <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))]</span>
<span class="c1"># identify SNPs that need flipping &amp; complement</span>
com.flip <span class="o">&lt;-</span> info<span class="p">[</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A2 <span class="o">&amp;</span>
                     <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">)</span> <span class="o">==</span> A1<span class="p">,</span> SNP<span class="p">]</span>
<span class="c1"># Now update the bim file</span>
bim<span class="p">[</span>SNP <span class="o">%in%</span> com.flip<span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span> <span class="o">:=</span>
        <span class="kt">list</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">),</span>
             <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))]</span>
<span class="c1"># And update the info structure</span>
info<span class="p">[</span>SNP <span class="o">%in%</span> com.flip<span class="p">,</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;B.A1&quot;</span><span class="p">,</span> <span class="s">&quot;B.A2&quot;</span><span class="p">)</span> <span class="o">:=</span>
        <span class="kt">list</span><span class="p">(</span><span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">),</span>
             <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))]</span>
<span class="c1"># Obtain list of SNPs that require flipping</span>
flip <span class="o">&lt;-</span> info<span class="p">[</span>B.A1<span class="o">==</span>A2 <span class="o">&amp;</span> B.A2<span class="o">==</span>A1<span class="p">]</span>
<span class="c1"># Now generate file for PLINK </span>
fwrite<span class="p">(</span>flip<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;SNP&quot;</span><span class="p">,</span> <span class="s">&quot;A1&quot;</span><span class="p">)],</span> <span class="s">&quot;EUR.update.a1&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="c1"># Write the updated bim file</span>
fwrite<span class="p">(</span>bim<span class="p">,</span> <span class="s">&quot;EUR.QC.adj.bim&quot;</span><span class="p">,</span> col.names<span class="o">=</span><span class="bp">F</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="c1"># We can then remove all mismatch SNPs</span>
matched <span class="o">&lt;-</span> info<span class="p">[(</span>A1 <span class="o">==</span> B.A1 <span class="o">&amp;</span> A2 <span class="o">==</span> B.A2<span class="p">)</span> <span class="o">|</span>
                    <span class="p">(</span>A1 <span class="o">==</span> B.A2 <span class="o">&amp;</span> A2 <span class="o">==</span> B.A1<span class="p">)</span> <span class="o">|</span>
                    <span class="p">(</span>A1 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">)</span> <span class="o">&amp;</span>
                         A2 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">))</span> <span class="o">|</span>
                    <span class="p">(</span>A1 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A2<span class="p">,</span> complement<span class="p">)</span> <span class="o">&amp;</span>
                         A2 <span class="o">==</span> <span class="kp">sapply</span><span class="p">(</span>B.A1<span class="p">,</span> complement<span class="p">))]</span>
mismatch <span class="o">&lt;-</span> bim<span class="p">[</span><span class="o">!</span>SNP<span class="o">%in%</span>matched<span class="o">$</span>SNP<span class="p">,</span> SNP<span class="p">]</span>
write.table<span class="p">(</span>mismatch<span class="p">,</span> <span class="s">&quot;EUR.mismatch&quot;</span><span class="p">,</span> quote<span class="o">=</span><span class="bp">F</span><span class="p">,</span> row.names<span class="o">=</span><span class="bp">F</span><span class="p">,</span> col.names<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</pre></div></div>
</div>
<p>The above script will generate three files: <strong>EUR.QC.adj.bim</strong>, <strong>EUR.update.a1</strong> and <strong>EUR.mismatch</strong>. </p>
<p>2. Replace <strong>EUR.QC.bim</strong> with <strong>EUR.QC.adj.bim</strong>:</p>
<div class="codehilite"><pre><span></span><span class="c1"># Make a back up</span>
mv EUR.bim EUR.bim.bk
ln -s EUR.QC.adj.bim EUR.bim
</pre></div>

<p>3. Generate a new genotype file with the alleles flipped so that the alleles in the base and target data match:
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR <span class="se">\</span>
    --a1-allele EUR.update.a1 <span class="se">\</span>
    --make-bed <span class="se">\</span>
    --keep EUR.QC.valid <span class="se">\</span>
    --exclude EUR.mismatch <span class="se">\</span>
    --out EUR.QC.flipped
</pre></div></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Most PRS software will perform flipping automatically, thus this step is usually not required.</p>
</div>
<h1 id="35-duplicate-snps"># Duplicate SNPs<a class="headerlink" href="#35-duplicate-snps" title="Permanent link">&para;</a></h1>
<p>Make sure to remove any duplicate SNPs in your target data (these target data were simulated and so include no duplicated SNPs)</p>
<h1 id="35-sample-overlap"># Sample overlap<a class="headerlink" href="#35-sample-overlap" title="Permanent link">&para;</a></h1>
<p>Since the target data were simulated there are no overlapping samples between the base and target data here (see the relevant section of [the paper] (https://doi.org/10.1101/416545) for discussion of the importance of avoiding sample overlap). </p>
<h1 id="35-relatedness"># Relatedness<a class="headerlink" href="#35-relatedness" title="Permanent link">&para;</a></h1>
<p>Closely related individuals in the target data may lead to overfitted results, limiting the generalisability of the results. </p>
<p>Before calculating the relatedness, pruning should be performed (see <a href="./#35-standard-gwas-qc">here</a>).
Individuals that have a first or second degree relative in the sample (<span class="arithmatex">\(\text{pi-hat} &gt; 0.125\)</span>) can be removed with the following command:</p>
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC.flipped <span class="se">\</span>
    --extract EUR.QC.prune.in <span class="se">\</span>
    --rel-cutoff <span class="m">0</span>.125 <span class="se">\</span>
    --out EUR.QC
</pre></div>

<details class="note"><summary>How many related samples were excluded?</summary><ul>
<li><code>2</code> samples were excluded</li>
</ul>
</details>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A greedy algorithm is used to remove closely related individuals in a way that optimises the size of the sample retained.                However, the algorithm is dependent on the random seed used, which can generate different results. Therefore, to reproduce
the same result, you will need to specify the same random seed. </p>
<p>PLINK's algorithm for removing related individuals does not account for the phenotype under study. 
To minimize the removal of cases of a disease, the following algorithm can be used instead: 
<a href="https://github.com/choishingwan/GreedyRelated">GreedyRelated</a>.</p>
</div>
<h1 id="generate-final-qced-target-data-file">Generate final QC'ed target data file<a class="headerlink" href="#generate-final-qced-target-data-file" title="Permanent link">&para;</a></h1>
<p>After performing the full analysis, you can generate a QC'ed data set with the following command:
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC.flipped <span class="se">\</span>
    --make-bed <span class="se">\</span>
    --out EUR.QC <span class="se">\</span>
    --extract EUR.QC.snplist
</pre></div></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../plink/" class="btn btn-neutral float-right" title="PLINK">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../base/" class="btn btn-neutral" title="1. QC of Summary Statistics"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../base/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../plink/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
