<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>PLINK - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "PLINK";
    var mkdocs_page_input_path = "plink.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../base/">1. QC of Base Data</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">2. QC of Target Data</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Calculating and analysing PRS</span>
    <ul class="subnav">
                <li class=" current">
                    
    <a class="current" href="./">PLINK</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#background">Background</a></li>
    

    <li class="toctree-l3"><a href="#required-data">Required Data</a></li>
    

    <li class="toctree-l3"><a href="#update-effect-size">Update Effect Size</a></li>
    

    <li class="toctree-l3"><a href="#clumping">Clumping</a></li>
    

    <li class="toctree-l3"><a href="#generate-prs">Generate PRS</a></li>
    

    <li class="toctree-l3"><a href="#accounting-for-population-stratification">Accounting for Population Stratification</a></li>
    

    <li class="toctree-l3"><a href="#finding-the-best-fit-prs">Finding the "best-fit" PRS</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class="">
                    
    <a class="" href="../ldpred/">LDpred</a>
                </li>
                <li class="">
                    
    <a class="" href="../lassosum/">lassosum</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../plink_visual/">4. Visualizing PRS Results</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>3. Calculating and analysing PRS &raquo;</li>
        
      
    
    <li>PLINK</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/plink.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h1>
<p>In this section of the tutorial you will use four different software programs to compute PRS from the base and target data that you QC'ed in the previous two sections. On this page, you will compute PRS using the popular genetic analyses tool <code>plink</code> - while <code>plink</code> is not a dedicated PRS software, each of the steps required to compute PRS using the C+T standard approach can be performed in <code>plink</code> and carrying out this multi-step process can be a good way to learn the processes involved in computing PRS (which are typically performed automatically by PRS software). </p>
<h1 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h1>
<p>In the previous sections, we have generated the following files:</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.valid</strong></td>
<td align="center">This file contains the samples that passed all the QC</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.covariate</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
</tbody>
</table>
<h1 id="update-effect-size">Update Effect Size<a class="headerlink" href="#update-effect-size" title="Permanent link">&para;</a></h1>
<p>When the effect size relates to disease risk and is thus given as an odds ratio (OR), rather than BETA (for continuous traits), then the PRS is computed as a product of ORs. To simplify this calculation, we usually take the natural logarithm of the OR so that the PRS can be computed using a simple summation instead (which can be back-transformed afterwards). 
We can obtain the transformed summary statistics with <code>R</code>:</p>
<div class="superfences-tabs">
<input name="__tabs_1" type="radio" id="__tab_1_0" checked="checked" />
<label for="__tab_1_0">Without data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>dat <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="kp">gzfile</span><span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">),</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
dat<span class="o">$</span>OR <span class="o">&lt;-</span> <span class="kp">log</span><span class="p">(</span>dat<span class="o">$</span>OR<span class="p">)</span>
write.table<span class="p">(</span>dat<span class="p">,</span> <span class="s">&quot;Height.QC.Transformed&quot;</span><span class="p">,</span> quote<span class="o">=</span><span class="bp">F</span><span class="p">,</span> row.names<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="kp">q</span><span class="p">()</span> <span class="c1"># exit R</span>
</pre></div></div>
<input name="__tabs_1" type="radio" id="__tab_1_1" />
<label for="__tab_1_1">With data.table</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span><span class="kn">library</span><span class="p">(</span>data.table<span class="p">)</span>
dat <span class="o">&lt;-</span> fread<span class="p">(</span><span class="s">&quot;Height.QC.gz&quot;</span><span class="p">)</span>
fwrite<span class="p">(</span>dat<span class="p">[,</span>OR<span class="o">:=</span><span class="kp">log</span><span class="p">(</span>OR<span class="p">)],</span> <span class="s">&quot;Height.QC.Transformed&quot;</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="kp">q</span><span class="p">()</span> <span class="c1"># exit R</span>
</pre></div></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>It may be tempting to perform the log transofrmation using <code>awk</code>.
However, due to rounding of values performed in <code>awk</code>, less accurate results
may be obtained. Therefore, we recommend performing the transformation in <code>R</code> or allow the PRS software to perform the transformation directly.</p>
</div>
<h1 id="clumping">Clumping<a class="headerlink" href="#clumping" title="Permanent link">&para;</a></h1>
<p>Linkage disequilibrium, which corresponds to the correlation between the genotypes of genetic variants across the genome, makes identifying the contribution from causal independent genetic variants extremely challenging. One way of approximately capturing the right level of causal signal is to perform clumping, which removes SNPs in such a way that only weakly correlated SNPs are retained but preferentially retaining the SNPs most associated with the phenotype under study. Clumping can be performed using the following command in <code>plink</code>: </p>
<div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC <span class="se">\</span>
    --clump-p1 <span class="m">1</span> <span class="se">\</span>
    --clump-r2 <span class="m">0</span>.1 <span class="se">\</span>
    --clump-kb <span class="m">250</span> <span class="se">\</span>
    --clump Height.QC.Transformed <span class="se">\</span>
    --clump-snp-field SNP <span class="se">\</span>
    --clump-field P <span class="se">\</span>
    --out EUR
</pre></div>

<p>Each of the new parameters corresponds to the following</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">clump-p1</td>
<td align="center">1</td>
<td align="left">P-value threshold for a SNP to be included as an index SNP. 1 is selected such that all SNPs are include for clumping</td>
</tr>
<tr>
<td align="center">clump-r2</td>
<td align="center">0.1</td>
<td align="left">SNPs having <span class="arithmatex">\(r^2\)</span> higher than 0.1 with the index SNPs will be removed</td>
</tr>
<tr>
<td align="center">clump-kb</td>
<td align="center">250</td>
<td align="left">SNPs within 250k of the index SNP are considered for clumping</td>
</tr>
<tr>
<td align="center">clump</td>
<td align="center">Height.QC.Transformed</td>
<td align="left">Base data (summary statistic) file containing the P-value information</td>
</tr>
<tr>
<td align="center">clump-snp-field</td>
<td align="center">SNP</td>
<td align="left">Specifies that the column <code>SNP</code> contains the SNP IDs</td>
</tr>
<tr>
<td align="center">clump-field</td>
<td align="center">P</td>
<td align="left">Specifies that the column <code>P</code> contains the P-value information</td>
</tr>
</tbody>
</table>
<p>A more detailed description of the clumping process can be found <a href="https://www.cog-genomics.org/plink/1.9/postproc#clump">here</a></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <span class="arithmatex">\(r^2\)</span> values computed by <code>--clump</code> are based on maximum likelihood haplotype frequency estimates</p>
</div>
<p>This will generate <strong>EUR.clumped</strong>, containing the index SNPs after clumping is performed.
We can extract the index SNP ID by performing the following command:</p>
<div class="codehilite"><pre><span></span>awk <span class="s1">&#39;NR!=1{print $3}&#39;</span> EUR.clumped &gt;  EUR.valid.snp
</pre></div>

<blockquote>
<p><code>$3</code> because the third column contains the SNP ID</p>
</blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If your target data are small (e.g. N &lt; 500) then you can use the 1000 Genomes Project samples for the LD calculation.
Make sure to use the population that most closely reflects represents the base sample.</p>
</div>
<h1 id="generate-prs">Generate PRS<a class="headerlink" href="#generate-prs" title="Permanent link">&para;</a></h1>
<p><code>plink</code> provides a convenient function <code>--score</code> and <code>--q-score-range</code> for calculating polygenic scores.</p>
<p>We will need three files:</p>
<ol>
<li>The base data file: <strong>Height.QC.Transformed</strong></li>
<li>A file containing SNP IDs and their corresponding P-values (<code>$1</code> because SNP ID is located in the first column; <code>$8</code> because the P-value is located in the eighth column)
<div class="codehilite"><pre><span></span>awk <span class="s1">&#39;{print $1,$8}&#39;</span> Height.QC.Transformed &gt; SNP.pvalue
</pre></div></li>
<li>A file containing the different P-value thresholds for inclusion of SNPs in the PRS. Here calculate PRS corresponding to a few thresholds for illustration purposes:
<div class="codehilite"><pre><span></span><span class="nb">echo</span> <span class="s2">&quot;0.001 0 0.001&quot;</span> &gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.05 0 0.05&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.1 0 0.1&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.2 0 0.2&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.3 0 0.3&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.4 0 0.4&quot;</span> &gt;&gt; range_list
<span class="nb">echo</span> <span class="s2">&quot;0.5 0 0.5&quot;</span> &gt;&gt; range_list
</pre></div>
The format of the <strong>range_list</strong> file should be as follows:</li>
</ol>
<table>
<thead>
<tr>
<th align="center">Name of Threshold</th>
<th align="center">Lower bound</th>
<th align="center">Upper Bound</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The threshold boundaries are inclusive. For example, for the <code>0.05</code> threshold, we include all SNPs with P-value from 
<code>0</code> to <code>0.05</code>, <strong>including</strong> any SNPs with P-value equal to <code>0.05</code>.</p>
</div>
<p>We can then calculate the PRS with the following <code>plink</code> command:</p>
<p><div class="codehilite"><pre><span></span>plink <span class="se">\</span>
    --bfile EUR.QC <span class="se">\</span>
    --score Height.QC.Transformed <span class="m">1</span> <span class="m">4</span> <span class="m">11</span> header <span class="se">\</span>
    --q-score-range range_list SNP.pvalue <span class="se">\</span>
    --extract EUR.valid.snp <span class="se">\</span>
    --out EUR
</pre></div>
The meaning of the new parameters are as follows:</p>
<table>
<thead>
<tr>
<th align="center">Paramter</th>
<th align="center">Value</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">score</td>
<td align="center">Height.QC.Transformed 1 4 11 header</td>
<td align="left">We read from the <strong>Height.QC.Transformed</strong> file, assuming that the <code>1</code>st column is the SNP ID; <code>4</code>th column is the effective allele information; the <code>11</code>th column is the effect size estimate; and that the file contains a <code>header</code></td>
</tr>
<tr>
<td align="center">q-score-range</td>
<td align="center">range_test SNP.pvalue</td>
<td align="left">We want to calculate PRS based on the thresholds defined in <strong>range_test</strong>, where the threshold values (P-values) were stored in <strong>SNP.pvalue</strong></td>
</tr>
</tbody>
</table>
<p>The above command and range_list will generate 7 files:</p>
<ol>
<li>EUR.0.5.profile</li>
<li>EUR.0.4.profile</li>
<li>EUR.0.3.profile</li>
<li>EUR.0.2.profile</li>
<li>EUR.0.1.profile</li>
<li>EUR.0.05.profile</li>
<li>EUR.0.001.profile</li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The default formula for PRS calculation in PLINK is:</p>
<div class="arithmatex">\[
PRS_j =\frac{ \sum_i^NS_i*G_{ij}}{P*M_j}
\]</div>
<p>where the effect size of SNP <span class="arithmatex">\(i\)</span> is <span class="arithmatex">\(S_i\)</span>;  the number of effect alleles observed in sample <span class="arithmatex">\(j\)</span> is <span class="arithmatex">\(G_{ij}\)</span>; the ploidy of the sample is <span class="arithmatex">\(P\)</span> (is generally 2 for humans); the number of samples included in the PRS is <span class="arithmatex">\(N\)</span>; and the number of non-missing SNPs observed in sample <span class="arithmatex">\(j\)</span> is <span class="arithmatex">\(M_j\)</span>. If the sample has a missing genotype for SNP <span class="arithmatex">\(i\)</span>, then the population minor allele frequency multiplied by the ploidy (<span class="arithmatex">\(MAF_i*P\)</span>) is used instead of <span class="arithmatex">\(G_{ij}\)</span>.</p>
</div>
<h1 id="accounting-for-population-stratification">Accounting for Population Stratification<a class="headerlink" href="#accounting-for-population-stratification" title="Permanent link">&para;</a></h1>
<p>Population structure is the principal source of confounding in GWAS and is usually accounted for by incorporating principal components (PCs) as covariates. We can incorporate PCs into our PRS analysis to account for population stratification.</p>
<p>Again, we can calculate the PCs using <code>plink</code>: 
<div class="codehilite"><pre><span></span><span class="c1"># First, we need to perform prunning</span>
plink <span class="se">\</span>
    --bfile EUR.QC <span class="se">\</span>
    --indep-pairwise <span class="m">200</span> <span class="m">50</span> <span class="m">0</span>.25 <span class="se">\</span>
    --out EUR
<span class="c1"># Then we calculate the first 6 PCs</span>
plink <span class="se">\</span>
    --bfile EUR.QC <span class="se">\</span>
    --extract EUR.prune.in <span class="se">\</span>
    --pca <span class="m">6</span> <span class="se">\</span>
    --out EUR
</pre></div></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>One way to select the appropriate number of PCs is to perform GWAS on the phenotype under study with different numbers of PCs.
<a href="https://github.com/bulik/ldsc">LDSC</a> analysis can then be performed on the set of GWAS summary statistics and the GWAS that used the number of PCs that gave an LDSC intercept closest to 1 should correspond to that for which population structure was most accurately controlled for. </p>
</div>
<p>Here the PCs have been stored in the <strong>EUR.eigenvec</strong> file and can be used as covariates in the regression model to account for population stratification.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If the base and target samples are collected from different worldwide populations then the results from the PRS analysis may be biased (see Section 3.4 of our papper).</p>
</div>
<h1 id="finding-the-best-fit-prs">Finding the "best-fit" PRS<a class="headerlink" href="#finding-the-best-fit-prs" title="Permanent link">&para;</a></h1>
<p>The P-value threshold that provides the "best-fit" PRS under the C+T method is usually unknown. 
To approximate the "best-fit" PRS, we can perform a regression between PRS calculated at a range of P-value thresholds and then select the PRS that explains the highest phenotypic variance (please see Section 4.6 of our paper on overfitting issues). 
This can be achieved using <code>R</code> as follows:</p>
<div class="superfences-tabs">
<input name="__tabs_2" type="radio" id="__tab_2_0" checked="checked" />
<label for="__tab_2_0">detail</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>p.threshold <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="m">0.001</span><span class="p">,</span><span class="m">0.05</span><span class="p">,</span><span class="m">0.1</span><span class="p">,</span><span class="m">0.2</span><span class="p">,</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.4</span><span class="p">,</span><span class="m">0.5</span><span class="p">)</span>
<span class="c1"># Read in the phenotype file </span>
phenotype <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.height&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
<span class="c1"># Read in the PCs</span>
pcs <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.eigenvec&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="c1"># The default output from plink does not include a header</span>
<span class="c1"># To make things simple, we will add the appropriate headers</span>
<span class="c1"># (1:6 because there are 6 PCs)</span>
<span class="kp">colnames</span><span class="p">(</span>pcs<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">))</span> 
<span class="c1"># Read in the covariates (here, it is sex)</span>
covariate <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.covariate&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
<span class="c1"># Now merge the files</span>
pheno <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span><span class="kp">merge</span><span class="p">(</span>phenotype<span class="p">,</span> covariate<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">)),</span> pcs<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">))</span>
<span class="c1"># We can then calculate the null model (model with PRS) using a linear regression </span>
<span class="c1"># (as height is quantitative)</span>
null.model <span class="o">&lt;-</span> lm<span class="p">(</span>Height<span class="o">~</span><span class="m">.</span><span class="p">,</span> data<span class="o">=</span>pheno<span class="p">[,</span><span class="o">!</span><span class="kp">colnames</span><span class="p">(</span>pheno<span class="p">)</span><span class="o">%in%</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)])</span>
<span class="c1"># And the R2 of the null model is </span>
null.r2 <span class="o">&lt;-</span> <span class="kp">summary</span><span class="p">(</span>null.model<span class="p">)</span><span class="o">$</span>r.squared
prs.result <span class="o">&lt;-</span> <span class="kc">NULL</span>
<span class="kr">for</span><span class="p">(</span>i <span class="kr">in</span> p.threshold<span class="p">){</span>
    <span class="c1"># Go through each p-value threshold</span>
    prs <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;EUR.&quot;</span><span class="p">,</span>i<span class="p">,</span><span class="s">&quot;.profile&quot;</span><span class="p">),</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
    <span class="c1"># Merge the prs with the phenotype matrix</span>
    <span class="c1"># We only want the FID, IID and PRS from the PRS file, therefore we only select the </span>
    <span class="c1"># relevant columns</span>
    pheno.prs <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>pheno<span class="p">,</span> prs<span class="p">[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="s">&quot;SCORE&quot;</span><span class="p">)],</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">))</span>
    <span class="c1"># Now perform a linear regression on Height with PRS and the covariates</span>
    <span class="c1"># ignoring the FID and IID from our model</span>
    model <span class="o">&lt;-</span> lm<span class="p">(</span>Height<span class="o">~</span><span class="m">.</span><span class="p">,</span> data<span class="o">=</span>pheno.prs<span class="p">[,</span><span class="o">!</span><span class="kp">colnames</span><span class="p">(</span>pheno.prs<span class="p">)</span><span class="o">%in%</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)])</span>
    <span class="c1"># model R2 is obtained as </span>
    model.r2 <span class="o">&lt;-</span> <span class="kp">summary</span><span class="p">(</span>model<span class="p">)</span><span class="o">$</span>r.squared
    <span class="c1"># R2 of PRS is simply calculated as the model R2 minus the null R2</span>
    prs.r2 <span class="o">&lt;-</span> model.r2<span class="o">-</span>null.r2
    <span class="c1"># We can also obtain the coeffcient and p-value of association of PRS as follow</span>
    prs.coef <span class="o">&lt;-</span> <span class="kp">summary</span><span class="p">(</span>model<span class="p">)</span><span class="o">$</span>coeff<span class="p">[</span><span class="s">&quot;SCORE&quot;</span><span class="p">,]</span>
    prs.beta <span class="o">&lt;-</span> <span class="kp">as.numeric</span><span class="p">(</span>prs.coef<span class="p">[</span><span class="m">1</span><span class="p">])</span>
    prs.se <span class="o">&lt;-</span> <span class="kp">as.numeric</span><span class="p">(</span>prs.coef<span class="p">[</span><span class="m">2</span><span class="p">])</span>
    prs.p <span class="o">&lt;-</span> <span class="kp">as.numeric</span><span class="p">(</span>prs.coef<span class="p">[</span><span class="m">4</span><span class="p">])</span>
    <span class="c1"># We can then store the results</span>
    prs.result <span class="o">&lt;-</span> <span class="kp">rbind</span><span class="p">(</span>prs.result<span class="p">,</span> <span class="kt">data.frame</span><span class="p">(</span>Threshold<span class="o">=</span>i<span class="p">,</span> R2<span class="o">=</span>prs.r2<span class="p">,</span> P<span class="o">=</span>prs.p<span class="p">,</span> BETA<span class="o">=</span>prs.beta<span class="p">,</span>SE<span class="o">=</span>prs.se<span class="p">))</span>
<span class="p">}</span>
<span class="c1"># Best result is:</span>
prs.result<span class="p">[</span><span class="kp">which.max</span><span class="p">(</span>prs.result<span class="o">$</span>R2<span class="p">),]</span>
<span class="kp">q</span><span class="p">()</span> <span class="c1"># exit R</span>
</pre></div></div>
<input name="__tabs_2" type="radio" id="__tab_2_1" />
<label for="__tab_2_1">quick</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>p.threshold <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="m">0.001</span><span class="p">,</span><span class="m">0.05</span><span class="p">,</span><span class="m">0.1</span><span class="p">,</span><span class="m">0.2</span><span class="p">,</span><span class="m">0.3</span><span class="p">,</span><span class="m">0.4</span><span class="p">,</span><span class="m">0.5</span><span class="p">)</span>
phenotype <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.height&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
pcs <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.eigenvec&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="kp">colnames</span><span class="p">(</span>pcs<span class="p">)</span> <span class="o">&lt;-</span> <span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;PC&quot;</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">6</span><span class="p">))</span> 
covariate <span class="o">&lt;-</span> read.table<span class="p">(</span><span class="s">&quot;EUR.covariate&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)</span>
pheno <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span><span class="kp">merge</span><span class="p">(</span>phenotype<span class="p">,</span> covariate<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">)),</span> pcs<span class="p">,</span> by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">))</span>
null.r2 <span class="o">&lt;-</span> <span class="kp">summary</span><span class="p">(</span>lm<span class="p">(</span>Height<span class="o">~</span><span class="m">.</span><span class="p">,</span> data<span class="o">=</span>pheno<span class="p">[,</span><span class="o">!</span><span class="kp">colnames</span><span class="p">(</span>pheno<span class="p">)</span><span class="o">%in%</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)]))</span><span class="o">$</span>r.squared
prs.result <span class="o">&lt;-</span> <span class="kc">NULL</span>
<span class="kr">for</span><span class="p">(</span>i <span class="kr">in</span> p.threshold<span class="p">){</span>
    pheno.prs <span class="o">&lt;-</span> <span class="kp">merge</span><span class="p">(</span>pheno<span class="p">,</span> 
                        read.table<span class="p">(</span><span class="kp">paste0</span><span class="p">(</span><span class="s">&quot;EUR.&quot;</span><span class="p">,</span>i<span class="p">,</span><span class="s">&quot;.profile&quot;</span><span class="p">),</span> header<span class="o">=</span><span class="bp">T</span><span class="p">)[,</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">,</span> <span class="s">&quot;SCORE&quot;</span><span class="p">)],</span>
                        by<span class="o">=</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span> <span class="s">&quot;IID&quot;</span><span class="p">))</span>
    model <span class="o">&lt;-</span> <span class="kp">summary</span><span class="p">(</span>lm<span class="p">(</span>Height<span class="o">~</span><span class="m">.</span><span class="p">,</span> data<span class="o">=</span>pheno.prs<span class="p">[,</span><span class="o">!</span><span class="kp">colnames</span><span class="p">(</span>pheno.prs<span class="p">)</span><span class="o">%in%</span><span class="kt">c</span><span class="p">(</span><span class="s">&quot;FID&quot;</span><span class="p">,</span><span class="s">&quot;IID&quot;</span><span class="p">)]))</span>
    model.r2 <span class="o">&lt;-</span> model<span class="o">$</span>r.squared
    prs.r2 <span class="o">&lt;-</span> model.r2<span class="o">-</span>null.r2
    prs.coef <span class="o">&lt;-</span> model<span class="o">$</span>coeff<span class="p">[</span><span class="s">&quot;SCORE&quot;</span><span class="p">,]</span>
    prs.result <span class="o">&lt;-</span> <span class="kp">rbind</span><span class="p">(</span>prs.result<span class="p">,</span> 
        <span class="kt">data.frame</span><span class="p">(</span>Threshold<span class="o">=</span>i<span class="p">,</span> R2<span class="o">=</span>prs.r2<span class="p">,</span> 
                    P<span class="o">=</span><span class="kp">as.numeric</span><span class="p">(</span>prs.coef<span class="p">[</span><span class="m">4</span><span class="p">]),</span> 
                    BETA<span class="o">=</span><span class="kp">as.numeric</span><span class="p">(</span>prs.coef<span class="p">[</span><span class="m">1</span><span class="p">]),</span>
                    SE<span class="o">=</span><span class="kp">as.numeric</span><span class="p">(</span>prs.coef<span class="p">[</span><span class="m">2</span><span class="p">])))</span>
<span class="p">}</span>
<span class="kp">print</span><span class="p">(</span>prs.result<span class="p">[</span><span class="kp">which.max</span><span class="p">(</span>prs.result<span class="o">$</span>R2<span class="p">),])</span>
<span class="kp">q</span><span class="p">()</span> <span class="c1"># exit R</span>
</pre></div></div>
</div>
<details class="note"><summary>Which P-value threshold generates the "best-fit" PRS?</summary><p>0.2 </p>
</details>
<details class="note"><summary>How much phenotypic variation does the "best-fit" PRS explain?</summary><p>0.04003232</p>
</details>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../prsice/" class="btn btn-neutral float-right" title="PRSice-2">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../target/" class="btn btn-neutral" title="2. QC of Target Data"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../target/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../prsice/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
