<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>1. QC of Base Data - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  <link href="../css/extra.css" rel="stylesheet">
  
  <script>
    // Current page data
    var mkdocs_page_name = "1. QC of Base Data";
    var mkdocs_page_input_path = "base.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">1. QC of Base Data</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#obtaining-the-base-data-file">Obtaining the base data file</a></li>
    

    <li class="toctree-l2"><a href="#reading-the-base-data-file">Reading the base data file</a></li>
    

    <li class="toctree-l2"><a href="#qc-checklist-base-data">QC checklist: Base data</a></li>
    

    <li class="toctree-l2"><a href="#35-heritability-check"># Heritability check</a></li>
    

    <li class="toctree-l2"><a href="#35-effect-allele"># Effect allele</a></li>
    

    <li class="toctree-l2"><a href="#35-file-transfer"># File transfer</a></li>
    

    <li class="toctree-l2"><a href="#35-genome-build"># Genome build</a></li>
    

    <li class="toctree-l2"><a href="#35-standard-gwas-qc"># Standard GWAS QC</a></li>
    

    <li class="toctree-l2"><a href="#35-ambiguous-snps"># Ambiguous SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-mismatching-genotypes"># Mismatching genotypes</a></li>
    

    <li class="toctree-l2"><a href="#35-duplicate-snps"># Duplicate SNPs</a></li>
    

    <li class="toctree-l2"><a href="#35-sex-chromosomes"># Sex chromosomes</a></li>
    

    <li class="toctree-l2"><a href="#35-sample-overlap"># Sample overlap</a></li>
    

    <li class="toctree-l2"><a href="#35-relatedness"># Relatedness</a></li>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../target/">2. QC of Target Data</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">3. Calculating and analysing PRS</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../plink/">PLINK</a>
                </li>
                <li class="">
                    
    <a class="" href="../prsice/">PRSice-2</a>
                </li>
                <li class="">
                    
    <a class="" href="../ldpred/">LDpred</a>
                </li>
                <li class="">
                    
    <a class="" href="../lassosum/">lassosum</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../plink_visual/">4. Visualizing PRS Results</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>1. QC of Base Data</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/base.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="obtaining-the-base-data-file">Obtaining the base data file<a class="headerlink" href="#obtaining-the-base-data-file" title="Permanent link">&para;</a></h1>
<p>The first step in Polygenic Risk Score (PRS) analyses is to generate or obtain the base data (GWAS summary statistics). Ideally these will correspond to the most powerful GWAS results available on the phenotype under study. In this example, we will use a modified version of the Height GWAS summary statistics generated by the <a href="https://portals.broadinstitute.org/collaboration/giant/index.php/GIANT_consortium_data_files#GWAS_Anthropometric_2014_Height">GIANT consortium</a>. You can download the summary statistic file <a href="https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/GIANT.height.gz">here</a> or you can use the following bash command:
<div class="codehilite"><pre><span></span>curl https://github.com/choishingwan/PRS-Tutorial/raw/master/resources/GIANT.height.gz -L -O
</pre></div></p>
<p>which will create a file called <strong>GIANT.height.gz</strong> in your working directory. </p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you download the summary statistics without using the bash command, and are using a MAC machine, the gz file will be decompressed automatically, resulting in a <strong>GIANT.height</strong> file instead. </p>
<p>To maintain consistency, we suggest compressing the <strong>GIANT.height</strong> file with 
<div class="codehilite"><pre><span></span>gzip GIANT.height
</pre></div>
before starting the tutorial</p>
</div>
<h1 id="reading-the-base-data-file">Reading the base data file<a class="headerlink" href="#reading-the-base-data-file" title="Permanent link">&para;</a></h1>
<p><strong>GIANT.height.gz</strong> is compressed. To read its content, you can type:</p>
<div class="codehilite"><pre><span></span>gunzip -c GIANT.height.gz <span class="p">|</span> head
</pre></div>

<p>which will display the first 10 lines of the file</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Working with compressed files reduces storage space requirements</p>
</div>
<p>The <strong>GIANT.height.gz</strong> file contains the following columns:</p>
<table>
<thead>
<tr>
<th align="center">SNP</th>
<th align="center">CHR</th>
<th align="center">BP</th>
<th align="center">A1</th>
<th align="center">A2</th>
<th align="center">MAF</th>
<th align="center">SE</th>
<th align="center">P</th>
<th align="center">N</th>
<th align="center">INFO</th>
<th align="center">OR</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">rs2073813</td>
<td align="center">1</td>
<td align="center">753541</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">0.125</td>
<td align="center">0.0083</td>
<td align="center">0.68</td>
<td align="center">69852</td>
<td align="center">0.866425782879888</td>
<td align="center">0.996605773454898</td>
</tr>
<tr>
<td align="center">rs12562034</td>
<td align="center">1</td>
<td align="center">768448</td>
<td align="center">A</td>
<td align="center">G</td>
<td align="center">0.092</td>
<td align="center">0.0088</td>
<td align="center">0.55</td>
<td align="center">88015</td>
<td align="center">0.917520990188678</td>
<td align="center">0.994714020220009</td>
</tr>
<tr>
<td align="center">rs2980319</td>
<td align="center">1</td>
<td align="center">777122</td>
<td align="center">A</td>
<td align="center">T</td>
<td align="center">0.125</td>
<td align="center">0.006</td>
<td align="center">0.65</td>
<td align="center">148975</td>
<td align="center">0.847126999058955</td>
<td align="center">0.997303641721713</td>
</tr>
</tbody>
</table>
<p>The column headers correspond to the following: </p>
<ul>
<li><strong>SNP</strong>: SNP ID, usually in the form of rs-ID</li>
<li><strong>CHR</strong>: The chromosome in which the SNP resides</li>
<li><strong>BP</strong>: Chromosomal co-ordinate of the SNP</li>
<li><strong>A1</strong>: The effect allele of the SNP</li>
<li><strong>A2</strong>: The non-effect allele of the SNP</li>
<li><strong>MAF</strong>: The minor allele frequency (MAF) of the SNP</li>
<li><strong>SE</strong>: The standard error (SE) of the effect size esimate</li>
<li><strong>P</strong>: The P-value of association between the SNP genotypes and the base phenotype</li>
<li><strong>N</strong>: Number of samples used to obtain the effect size estimate</li>
<li><strong>INFO</strong>: The imputation information score</li>
<li><strong>OR</strong>: The effect size estimate of the SNP, if the outcome is binary/case-control. If the outcome is continuous or treated as continuous then this will be the BETA</li>
</ul>
<h1 id="qc-checklist-base-data">QC checklist: Base data<a class="headerlink" href="#qc-checklist-base-data" title="Permanent link">&para;</a></h1>
<p>Below we perform QC on these base data according to the 'QC checklist' in the guide paper, which we recommend that users follow each time they perform a PRS analysis:</p>
<h1 id="35-heritability-check"># Heritability check<a class="headerlink" href="#35-heritability-check" title="Permanent link">&para;</a></h1>
<p>We recommend that PRS analyses are performed on base data with a chip-heritability estimate <span class="arithmatex">\(h_{snp}^{2} &gt; 0.05\)</span>. The chip-heritability of a GWAS can be estimated using e.g. LD Score Regression (LDSC). Our GIANT height GWAS data are known to have a chip-heritability much greater than 0.05 and so we can move on to the next QC step. </p>
<h1 id="35-effect-allele"># Effect allele<a class="headerlink" href="#35-effect-allele" title="Permanent link">&para;</a></h1>
<p>The GIANT consortium report which is the effect allele and which is the non-effect allele in their results, critical for PRS association results to be in the correction direction.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Some GWAS results files do not make clear which allele is the effect allele and which the non-effect allele.
If the incorrect assumption is made in computing the PRS, then the effect of the PRS in the target data will be in the wrong direction.</p>
<p>To avoid misleading conclusions the effect allele from the base (GWAS) data must be known.</p>
</div>
<h1 id="35-file-transfer"># File transfer<a class="headerlink" href="#35-file-transfer" title="Permanent link">&para;</a></h1>
<p>A common problem is that the downloaded base data file becomes 
corrupted during download, which can cause PRS software to crash 
or to produce errors in results. However, a <code>md5sum</code> hash is 
generally included in files so that file integrity can be checked. 
The following command performs this <code>md5sum</code> check: </p>
<div class="superfences-tabs">
<input name="__tabs_1" type="radio" id="__tab_1_0" checked="checked" />
<label for="__tab_1_0">Linux</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>md5sum GIANT.height.gz
</pre></div></div>
<input name="__tabs_1" type="radio" id="__tab_1_1" />
<label for="__tab_1_1">OS X</label>
<div class="superfences-content"><div class="codehilite"><pre><span></span>md5 GIANT.height.gz
</pre></div></div>
</div>
<p>if the file is intact, then <code>md5sum</code> generates a string of characters, which in this case should be: <code>80e48168416a2fdbe88d68cdfebd4ca2</code>. 
If a different string is generated, then the file is corrupted.</p>
<h1 id="35-genome-build"># Genome build<a class="headerlink" href="#35-genome-build" title="Permanent link">&para;</a></h1>
<p>These base data are on the same genome build as the target data that we will be using. You must check that your base and target data are on the same genome build, and if they are not then use a tool such as LiftOver to make the builds consistent across the data sets.</p>
<h1 id="35-standard-gwas-qc"># Standard GWAS QC<a class="headerlink" href="#35-standard-gwas-qc" title="Permanent link">&para;</a></h1>
<p>As described in the paper, both the base and target data should be subjected to the standard stringent QC steps performed in GWAS. 
If the base data have been obtained as summary statistics from a public source, then the typical QC steps that you will be able to perform on them are to filter the SNPs according to INFO score and MAF. 
SNPs with low minor allele frequency (MAF) or imputation information score (INFO) are more likely to generate false positive results due to their lower statistical power (and higher probability of genotyping errors in the case of low MAF). 
Therefore, SNPs with low MAF and INFO are typically removed before performing downstream analyses.
We recommend removing SNPs with MAF &lt; 1% and INFO &lt; 0.8 (with very large base sample sizes these thresholds could be reduced if sensitivity checks indicate reliable results).
These SNP filters can be acheived using the following code:</p>
<div class="codehilite"><pre><span></span>gunzip -c GIANT.height.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;NR==1 || ($6 &gt; 0.01) &amp;&amp; ($10 &gt; 0.8) {print}&#39;</span> <span class="p">|</span><span class="se">\</span>
gzip  &gt; Height.gz
</pre></div>

<p>The bash code above does the following:
1. Decompresses and reads the <strong>GIANT.height.gz</strong> file
2. Prints the header line (<code>NR==1</code>)
3. Prints any line with MAF above 0.01 (<code>$6</code> because the sixth column of the file contains the MAF information)
4. Prints any line with INFO above 0.8 (<code>$10</code> because the tenth column of the file contains the INFO information)
5. Compresses and writes the results to <strong>Height.gz</strong></p>
<h1 id="35-ambiguous-snps"># Ambiguous SNPs<a class="headerlink" href="#35-ambiguous-snps" title="Permanent link">&para;</a></h1>
<p>If the base and target data were generated using different genotyping chips and the chromosome strand (+/-) for either is unknown, then it is not possible to match ambiguous SNPs (i.e. those with complementary alleles, either C/G or A/T) across the data sets, because it will be unknown whether the base and target data are referring to the same allele or not. Ambiguous SNPs can be removed in the base data and then there will be no such SNPs in the subsequent analyses, since analyses are performed only on SNPs that overlap between the base and target data.</p>
<p>Nonambiguous SNPs can be retained using the following:
<div class="codehilite"><pre><span></span>gunzip -c Height.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;!( ($4==&quot;A&quot; &amp;&amp; $5==&quot;T&quot;) || \</span>
<span class="s1">        ($4==&quot;T&quot; &amp;&amp; $5==&quot;A&quot;) || \</span>
<span class="s1">        ($4==&quot;G&quot; &amp;&amp; $5==&quot;C&quot;) || \</span>
<span class="s1">        ($4==&quot;C&quot; &amp;&amp; $5==&quot;G&quot;)) {print}&#39;</span> <span class="p">|</span><span class="se">\</span>
    gzip &gt; Height.noambig.gz
</pre></div></p>
<details class="note"><summary>How many non-ambiguous SNPs were there?</summary><p>There are <code>608,811</code> ambiguous SNPs</p>
</details>
<h1 id="35-mismatching-genotypes"># Mismatching genotypes<a class="headerlink" href="#35-mismatching-genotypes" title="Permanent link">&para;</a></h1>
<p>If there is a non-ambiguous mismatch in allele coding between the base and target data sets, such as A/C in the base and G/T in the target data, then this can be resolved by ‘flipping’ the alleles in either data set to their complementary alleles. However, since we need the target data to know which SNPs have mismatching genotypes across the data sets, then we will perform this 'allele flipping' in the target data.</p>
<h1 id="35-duplicate-snps"># Duplicate SNPs<a class="headerlink" href="#35-duplicate-snps" title="Permanent link">&para;</a></h1>
<p>If an error has occurred in the generation of the base data then there may be duplicated SNPs in the base data file.
Most PRS software do not allow duplicated SNPs in the base data input and thus they should be removed, using a command such as the one below: </p>
<div class="codehilite"><pre><span></span>gunzip -c Height.noambig.gz <span class="p">|</span><span class="se">\</span>
awk <span class="s1">&#39;{ print $1}&#39;</span> <span class="p">|</span><span class="se">\</span>
sort <span class="p">|</span><span class="se">\</span>
uniq -d &gt; duplicated.snp
</pre></div>

<p>The above command does the following:</p>
<ol>
<li>Decompresses and reads the <strong>Height.noambig.gz</strong> file</li>
<li>Prints out the first column of the file (which contains the SNP ID; change <code>$1</code> to another number if the SNP ID is located in another column, e.g. <code>$3</code> if the SNP ID is located on the third column)</li>
<li>Sort the SNP IDs. This will put duplicated SNP IDs next to each other</li>
<li>Print out any duplicated SNP IDs using the uniq command and print them to the <em>duplicated.snp</em> file</li>
</ol>
<details class="note"><summary>How many duplicated SNPs are there?</summary><p>There are a total of <code>10</code> duplicated SNPs</p>
</details>
<p>Duplicated SNPs can then be removed using the <code>grep</code> command:
<div class="codehilite"><pre><span></span>gunzip -c Height.noambig.gz  <span class="p">|</span><span class="se">\</span>
grep -vf duplicated.snp <span class="p">|</span><span class="se">\</span>
gzip - &gt; Height.QC.gz
</pre></div></p>
<p>The above script does the following:</p>
<ol>
<li>Decompresses and reads the <strong>Height.noambig.gz</strong> file </li>
<li>Establishes whether any row contains entries observed in <code>duplicated.snp</code> and removes them if so</li>
<li>Compresses and writes the results to <strong>Height.QC.gz</strong></li>
</ol>
<h1 id="35-sex-chromosomes"># Sex chromosomes<a class="headerlink" href="#35-sex-chromosomes" title="Permanent link">&para;</a></h1>
<p>Previously performed QC on these data removed individuals with mismatching (inferred) biological and reported sex, while the sex chromosomes are not included. Please refer to the corresponding section in the paper for details relating QC performed in relation to the sex chromosomes. </p>
<h1 id="35-sample-overlap"># Sample overlap<a class="headerlink" href="#35-sample-overlap" title="Permanent link">&para;</a></h1>
<p>In this tutorial the target data are simulated and thus there must be no sample overlap. However, users should ensure that the possibility of sample overlap between the base and target data is minimised. </p>
<h1 id="35-relatedness"># Relatedness<a class="headerlink" href="#35-relatedness" title="Permanent link">&para;</a></h1>
<p>In this tutorial the target data are simulated and thus there must be no closely related individuals across the base and target data. However, users should ensure that the possibility of closely related individuals between the base and target data is minimised. </p>
<p>The <strong>Height.QC.gz</strong> base data are now ready for using in downstream analyses.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../target/" class="btn btn-neutral float-right" title="2. QC of Target Data">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Overview"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../target/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
